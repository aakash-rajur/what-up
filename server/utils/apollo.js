const {ApolloServer, gql} = require('apollo-server-express');
const getPublisher = require('./publisher');
const cookie = require('cookie');
const publisher = getPublisher();
const {
	TASKS_CHANGED,
	TASK_CANCELLED,
	SESSION_CHANGE,
	verifySession,
	getStats
} = require('./library');

const typeDefs = gql`
    enum TASK_STATUS {
        CREATED,
        COMPLETED,
        CANCELLED
    }

    type Task {
        id: String!,
        description: String!,
        status: TASK_STATUS
        created: String!,
        updated: String!
    }

    type Event {
        timestamp: String!,
        CREATED: Int,
        COMPLETED: Int,
        CANCELLED: Int,
        ALL: Int,
        additional: String
    }

    type Query {
        hello(name: String): String!,
        tasks(filter: String, timestamp: String): [Task]
    }

    type Mutation {
        hello(name: String): String!,
        add(description: String!): String!,
        remove(id: String!): String!,
        edit(id: String!, description: String!): String!
        update(id: String!, status: String!): String!,
        updateAll(filter:String!, status: String!): Int
    }

    type Subscription {
        TASKS_CHANGED: Event!,
        SESSION_CHANGE: Event!
    }
`;

function resolverGenerator(DB, postgres) {
	return {
		Query: {
			hello: (root, {name}) => {
				return `hello ${name || 'world'}`;
			},
			tasks: async (root, {filter, timestamp}, context) => {
				return postgres.getTasks(context.user, filter);
			}
		},
		Mutation: {
			hello: (root, {name}) => `not hello ${name || 'world'}`,
			add: async (root, {description}, context) => {
				let result = await postgres.addTask(context.user, description);
				publisher.notify(TASKS_CHANGED, await getStats(context.user));
				return result;
			},
			remove: async (root, {id}, context) => {
				let result = await postgres.updateTask(id, TASK_CANCELLED);
				publisher.notify(TASKS_CHANGED, await getStats(context.user));
				return result;
			},
			edit: async (root, {id, description}, context) => {
				let result = await postgres.editTask(id, description);
				publisher.notify(TASKS_CHANGED, await getStats(context.user));
				return result;
			},
			update: async (root, {id, status}, context) => {
				let result = await postgres.updateTask(id, status);
				publisher.notify(TASKS_CHANGED, await getStats(context.user));
				return result;
			},
			updateAll: async (root, {filter, status}, context) => {
				let result = await postgres.updateAllTasks(context.user, filter, status);
				publisher.notify(TASKS_CHANGED, await getStats(context.user));
				return result;
			}
		},
		Subscription: {
			[TASKS_CHANGED]: {
				subscribe: () => publisher.asyncIterator([TASKS_CHANGED])
			},
			[SESSION_CHANGE]: {
				subscribe: () => publisher.asyncIterator([SESSION_CHANGE])
			}
		}
	}
}

function getApolloServer(DB = {}, postgres) {
	return new ApolloServer({
		cors: false,
		typeDefs,
		resolvers: resolverGenerator(DB, postgres),
		subscriptions: {
			onConnect: (connectionParams, webSocket) => {
				let {remoteAddress, remotePort} = webSocket._socket;
				console.log(`websocket connected to ${remoteAddress}:${remotePort}`);
				let {session} = cookie.parse(webSocket.upgradeReq.headers.cookie);
				try {
					let user = verifySession(session);
					console.log(user);
					if (user) setImmediate(async () => {
						publisher.notify(TASKS_CHANGED, await getStats(user));
						publisher.notify(SESSION_CHANGE, {
							additional: JSON.stringify({
								message: `Your UserID is ${user}`
							})
						});
					});
				} catch (e) {
					console.error(e);
				}
			}
		},
		context: ({req}) => ({...req})
	});
}

module.exports = {
	getApolloServer
};
