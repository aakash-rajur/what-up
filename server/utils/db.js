require('dotenv').config();
const {NODE_ENV} = process.env;
const {Pool} = require('pg');

let instance = null;

function formatQuery(result, reduce = true) {
	let {
		rowCount,
		rows
	} = result;
	if (reduce && rowCount === 1) return rows[0];
	return rows;
}

class Postgres {
	constructor(uri = null) {
		if (!uri) throw new Error('uri not provided');
		this.uri = uri;
		this.pool = new Pool({
			connectionString: uri,
			ssl: NODE_ENV === 'production'
		});
	}
	
	checkConnection() {
		if (!this.uri) throw new Error('uri not provided');
		if (!this.pool) throw new Error('pool not instantiated');
	}
	
	disconnect() {
		this.checkConnection();
		return this.pool.end();
	}
	
	async addUser(hash) {
		this.checkConnection();
		return formatQuery(await this.pool.query(`select * from add_user('${hash}')`));
	}
	
	async deleteUser(hash) {
		this.checkConnection();
		await this.pool.query(`select * from delete_tasks('${hash}')`);
		return formatQuery(await this.pool.query(`select * from delete_user('${hash}')`));
	}
	
	async addTask(hash, description) {
		this.checkConnection();
		return formatQuery(await this.pool.query(`select * from add_task('${hash}', '${description}')`));
	}
	
	async updateTask(taskId, newStatus) {
		this.checkConnection();
		return formatQuery(await this.pool.query(`select * from update_task('${taskId}','${newStatus}')`));
	}
	
	async editTask(taskId, description) {
		this.checkConnection();
		return formatQuery(await this.pool.query(`select * from edit_task('${taskId}', '${description}')`));
	}
	
	async deleteTasks(user) {
		this.checkConnection();
		return formatQuery(await this.pool.query(`select * from delete_tasks('${user}')`));
	}
	
	async getTasks(hash, status) {
		this.checkConnection();
		return formatQuery(await this.pool.query(`select * from get_tasks('${hash}', '${status}')`), false);
	}
	
	async updateAllTasks(hash, selected, status) {
		this.checkConnection();
		return formatQuery(await this.pool.query(`select * from update_all_tasks('${hash}','${selected}','${status}')`));
	}
	
	async getStats(hash) {
		this.checkConnection();
		return formatQuery(await this.pool.query(`select * from get_stats('${hash}')`), false);
	}
	
	query(...args) {
		return this.pool.query(...args);
	}
}

function getDB(uri) {
	if (!instance){
		console.log('environment', NODE_ENV);
		instance = new Postgres(uri);
	}
	return instance;
}

module.exports = getDB;
