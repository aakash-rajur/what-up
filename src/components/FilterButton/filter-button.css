@import "../../utils/variables.css";

.filter {
    background: none;
    border: none;
    font-size: 0.8em;
    font-weight: 100;
    padding: 0.5em 2em 0.5em 1em;
    border-radius: 4px;
    cursor: pointer;
    color: white;
    transition: all 0.3s;
    outline: none;
    background: lightgray;
    box-shadow: 0 1px 3px rgba(0,0,0,0.12), 0 1px 2px rgba(0,0,0,0.24);
    position: relative;
    display: flex;
    align-items: center;
    justify-content: center;
}

.filter:hover {
    box-shadow: 0 3px 6px rgba(0,0,0,0.16), 0 3px 6px rgba(0,0,0,0.23);
    background: gray;
}

.filter[data-active="true"] {
    box-shadow: 0 3px 6px rgba(0,0,0,0.16), 0 3px 6px rgba(0,0,0,0.23);
    background: deepskyblue;
    color: white;
    cursor: not-allowed;
}


.badge {
    position: absolute;
    background: white;
    color: var(--accent-opaque);
    width: 16px;
    height: 16px;
    border-radius: 50%;
    right: 4px;
    top: calc(50% - 8px);
    font-size: 0.8em;
    display: flex;
    align-items: center;
    justify-content: center;
}

@media (pointer: coarse) {
    .filter {
        font-size: 1em;
    }
}
