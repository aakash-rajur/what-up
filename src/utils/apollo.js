import {InMemoryCache} from 'apollo-cache-inmemory';
import {ApolloClient} from 'apollo-client';
import {ApolloLink, split} from 'apollo-link';
import {onError} from 'apollo-link-error';
import {HttpLink} from 'apollo-link-http';
import {WebSocketLink} from 'apollo-link-ws';
import {getMainDefinition} from "apollo-utilities";
import gql from 'graphql-tag';
import {compose, graphql} from 'react-apollo';
import {
	API_URL,
	SESSION_CHANGE,
	TASK_ALL,
	TASK_CANCELLED,
	TASK_COMPLETED,
	TASK_CREATED,
	TASKS_CHANGED,
	WS_URL
} from "./constants";

const wsLink = new WebSocketLink({
	uri: WS_URL,
	options: {
		reconnect: true
	}
});

export default new ApolloClient({
	link: ApolloLink.from([
		onError(({graphQLErrors, networkError}) => {
			if (graphQLErrors)
				graphQLErrors.map(({message, locations, path}) =>
					console.log(
						`[GraphQL error]: Message: ${message}, Location: ${locations}, Path: ${path}`,
					),
				);
			if (networkError) console.log(`[Network error]: ${JSON.stringify(networkError)}`);
		}),
		split(({query}) => {
				const {kind, operation} = getMainDefinition(query);
				return kind === 'OperationDefinition' && operation === 'subscription';
			},
			wsLink,
			new HttpLink({
				uri: API_URL,
				credentials: 'include'
			}))
	]),
	cache: new InMemoryCache()
});

export const FETCH_TASKS = gql`
    query fetchTasks($filter: String!, $timestamp: String){
        tasks(filter: $filter, timestamp: $timestamp) {
            id,
            description,
            status,
            created,
            updated
        }
    }
`;

export const CANCEL_TASK = gql`
    mutation removeTask($id: String!) {
        remove(id: $id)
    }
`;

export const EDIT_TASK = gql`
    mutation editTask($id: String!, $description: String!) {
        edit(id: $id, description: $description)
    }
`;

export const UPDATE_TASK = gql`
    mutation updateTask($id: String!, $status: String!) {
        update(id: $id, status: $status)
    }
`;

export const UPDATE_ALL_TASKS = gql`
    mutation updateAll($filter: String!, $status: String!) {
        updateAll(filter: $filter,status: $status)
    }
`;

export const ADD_TASK = gql`
    mutation addTask($description: String!) {
        add(description: $description)
    }
`;

export const TASKS_UPDATED = gql`
    subscription onTasksChanged {
        TASKS_CHANGED {
            timestamp,
            CREATED,
            COMPLETED,
            CANCELLED,
            ALL
        }
    }
`;

export const SESSION_CHANGED = gql`
    subscription onSessionChanged {
        SESSION_CHANGE {
            additional
        }
    }
`;

export const withTaskMutations = compose(...[{
		mutation: CANCEL_TASK,
		prefix: 'cancel',
		args: ({id}) => ({id})
	}, {
		mutation: EDIT_TASK,
		prefix: 'edit',
		args: () => ({}),
		dynamic: true
	}, {
		mutation: UPDATE_TASK,
		prefix: 'update',
		args: ({id, status}) => ({
			id, status: status === TASK_CREATED ?
				TASK_COMPLETED : TASK_CREATED
		})
	}].map(({mutation, prefix, args, dynamic}) => graphql(mutation, {
		props: ({mutate}, result) => {
			return ({
				[`${prefix}Task`]: dynamic ? props => mutate({variables: {...props}}) : mutate,
				[`${prefix}Result`]: result ? result.data : null
			})
		},
		options: props => ({variables: args(props)})
	}))
);

export const withQueryTasks = graphql(FETCH_TASKS, {
	props: ({data = {}}, ...rest) => ({...data, ...rest})
});

export const withNewTaskAddition = graphql(ADD_TASK, {
	props: ({mutate}, result) => ({addNewTask: mutate, result}),
	options: ({newTask, onDone}) => ({
		variables: {description: newTask},
		onCompleted: onDone,
		onError: onDone
	})
});

export const withUpdateAll = graphql(UPDATE_ALL_TASKS, {
	props: ({mutate, result}) => ({updateAll: mutate, result}),
	options: ({filter, nextStatus}) => ({variables: {filter, status: nextStatus}})
});

const defaultTaskUpdated = {
	[TASKS_CHANGED]: {
		timestamp: "0",
		[TASK_ALL]: 0,
		[TASK_CANCELLED]: 0,
		[TASK_COMPLETED]: 0,
		[TASK_CREATED]: 0
	}
};

const defaultSession = {
	[SESSION_CHANGE]: null
};

export const withSessionAndTaskSubscription = compose(
	graphql(SESSION_CHANGED, {
		props: ({data: {[SESSION_CHANGE]: session} = defaultSession}) => ({
			session: session && JSON.parse(session.additional)
		}),
		options: () => ({shouldResubscribe: true})
	}),
	graphql(TASKS_UPDATED, {
		props: ({data: {[TASKS_CHANGED]: stats} = defaultTaskUpdated}) => ({...stats}),
		options: () => ({shouldResubscribe: true})
	}),
);
