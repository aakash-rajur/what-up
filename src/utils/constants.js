export const API_URL = process.env.REACT_APP_API_URL;
export const WS_URL = process.env.REACT_APP_SOCKET_URL;
export const TASKS_CHANGED = process.env.REACT_APP_TASKS_CHANGED;
export const TASK_CREATED = process.env.REACT_APP_TASK_CREATED;
export const TASK_COMPLETED = process.env.REACT_APP_TASK_COMPLETED;
export const TASK_CANCELLED = process.env.REACT_APP_TASK_CANCELLED;
export const TASK_ALL = process.env.REACT_APP_TASK_ALL;
export const SESSION_CHANGE = process.env.REACT_APP_SESSION_CHANGE;
export const TASK_TITLE = 'Task description\nDouble click to edit';
