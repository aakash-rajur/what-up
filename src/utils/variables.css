:root {
    --accent-color: rgba(175, 47, 47, 0.15);
    --accent-opaque: #cc9a9a;
    --app-width: 550px;
    --icon-size: 24px;
}

@media (max-width: 600px) {
    :root {
        --icon-size: 20px;
    }
}
