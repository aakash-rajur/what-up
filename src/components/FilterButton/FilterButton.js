import React, {Component} from 'react';
import PropTypes from 'prop-types';
import './filter-button.css'

class FilterButton extends Component {
    static propTypes = {
        filter: PropTypes.string,
        onClick: PropTypes.func,
        className: PropTypes.string,
        title: PropTypes.string,
        active: PropTypes.bool,
        stat: PropTypes.number,
    };

    static defaultProps = {
        active: false
    };

    constructor(props) {
        super(props);
        this.onClick = this.onClick.bind(this);
    }

    render() {
        let {
            title, active,
            filter, stat
        } = this.props;
        return (
            <button title={title} className={`filter ${(filter || '').toLowerCase()}`}
                    onClick={this.onClick} data-active={active} disabled={active}>
                {title}
                {stat && <span className="badge">{stat > 99 ? 'âˆž' : stat}</span>}
            </button>
        )
    }

    onClick() {
        let {onClick, filter} = this.props;
        onClick && onClick(filter);
    }
}

export default FilterButton;
