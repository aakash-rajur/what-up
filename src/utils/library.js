import React from "react";
import {Mutation} from 'react-apollo';
import {TASK_ALL, TASK_CANCELLED, TASK_COMPLETED, TASK_CREATED} from "./constants";

export function promiseSetState(context) {
	return newState => new Promise(resolve =>
		context.setState(newState, resolve)
	);
}

export function wait(delay, cb, ...args) {
	let timeout = null,
		promise = new Promise(resolve => {
			timeout = setTimeout(async () =>
				resolve(cb && await cb(...(args || []))),
				delay
			);
		});
	promise.cancel = () => clearTimeout(timeout);
	return promise;
}

export function renderMutations(config) {
	return config.map(({mutation, getVariables, renderComponent}, index) => {
		if (!(renderComponent instanceof Function)) return renderComponent;
		return (
			<Mutation key={index} mutation={mutation} variables={getVariables(mutation)}>
				{renderComponent}
			</Mutation>
		)
	});
}

export function getNextUpdateAllStatus(tasks = []) {
	let task = tasks.find(({status}) => status !== TASK_COMPLETED);
	return task ? TASK_COMPLETED : TASK_CREATED;
}

const DAY_MAP = ["Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat"];
const MONTH_MAP = ["Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"];

function padStart(string) {
	return typeof string === 'string' ? string.padStart(2, '0') :
		string.toString().padStart(2, '0');
}

export function getFormattedTimestamp(timestamp) {
	let date = new Date(timestamp);
	// noinspection JSCheckFunctionSignatures
	if (isNaN(date)) return 'Invalid Date';
	let hr = date.getHours(),
		suffix = 'A.M';
	if (hr === 0) {
		hr = 12;
		suffix = 'A.M';
	} else if (hr === 12) {
		suffix = 'P.M';
	} else hr %= 12;
	return `${padStart(hr)}: ${padStart(date.getMinutes())} ${suffix}, ${DAY_MAP[date.getDay()]} ${date.getDate()} ${MONTH_MAP[date.getMonth()]}`
}

export const FILTER_BUTTON_TEMPLATE = [{
	text: 'All',
	title: 'View All Tasks',
	filter: TASK_ALL
}, {
	text: 'Active',
	title: 'View Active Tasks',
	filter: TASK_CREATED
}, {
	text: 'Completed',
	title: 'View Completed Tasks',
	filter: TASK_COMPLETED
}, {
	text: 'Cancelled',
	title: 'View Cancelled Tasks',
	filter: TASK_CANCELLED
}];
