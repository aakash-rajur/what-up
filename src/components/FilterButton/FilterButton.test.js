import React from "react";
import {TASK_ALL} from "../../utils/constants";
import FilterButton from "./FilterButton";

describe("FilterButton with default props", () => {
  let active = false,
    component = shallow(<FilterButton />);

  it("should render", () => {
    expect(component.find("button.filter").exists()).toBe(true);
  });

  it(
    "dom tree integrity",
    checkDOM(component, [
      {
        selector: "button.filter",
        props: {
          title: "",
          "data-active": "false",
          disabled: false
        },
        children: [
          {
            selector: "span.badge",
            length: 1
          }
        ]
      },
      {
        selector: "span.badge",
        text: "0"
      }
    ])
  );

  matchSnapshot(component);
});

describe("FilterButton with props", () => {
  let props = {
      active: true,
      title: TASK_ALL,
      stat: 7,
      filter: "filter1",
      onClick: stub()
    },
    component = shallow(<FilterButton {...props} />);

  it("should render", () => {
    expect(component.find("button.filter").exists()).toBe(true);
  });

  it(
    "dom tree integrity",
    checkDOM(component, [
      {
        selector: "button.filter",
        props: {
          title: props.title,
          className: "filter filter1",
          "data-active": props.active,
          disabled: props.active
        },
        children: [
          {
            selector: "span.badge",
            length: 1
          }
        ]
      },
      {
        selector: "span.badge",
        text: props.stat.toString()
      }
    ])
  );

  it(
    "fires onClick",
    checkEvents(component, [
      {
        selector: "button.filter",
        event: "click",
        callback: props.onClick,
        cArgs: [props.filter]
      }
    ])
  );

  matchSnapshot(component);
});
