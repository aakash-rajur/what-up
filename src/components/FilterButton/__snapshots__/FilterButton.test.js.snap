// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterButton with default props match snapshot with default props 1`] = `
<button
  className="filter "
  data-active={false}
  disabled={false}
  onClick={[Function]}
  title=""
>
  <span
    className="badge"
  >
    0
  </span>
</button>
`;

exports[`FilterButton with props match snapshot with non-default props 1`] = `
<button
  className="filter filter1"
  data-active={true}
  disabled={true}
  onClick={[Function]}
  title="All"
>
  All
  <span
    className="badge"
  >
    7
  </span>
</button>
`;
